src/object/srv_obj_migrate.c

```diff
@@ -131,64 +130,47 @@ obj_tree_destroy(daos_handle_t btr_hdl)
 	return rc;
 }
 
-/* Create tree root by key_iov */
 static int
-tree_cache_create_internal(daos_handle_t toh, unsigned int tree_class,
-			   d_iov_t *key_iov, struct tree_cache_root **rootp)
+container_tree_create(daos_handle_t toh, uuid_t uuid,
+		      struct tree_cache_root **rootp)
 {
+	d_iov_t			key_iov;
 	d_iov_t			val_iov;
 	struct umem_attr	uma;
-	struct tree_cache_root	root;
-	struct btr_root		*broot;
+	struct tree_cache_root	root = { 0 };
+	struct tree_cache_root	*tmp_root;
 	int			rc;
 
-	D_ALLOC_PTR(broot);
-	if (broot == NULL)
-		return -DER_NOMEM;
+	d_iov_set(&key_iov, uuid, sizeof(uuid_t));
+	d_iov_set(&val_iov, &root, sizeof(root));
+	rc = dbtree_update(toh, &key_iov, &val_iov);
+	if (rc)
+		D_GOTO(out, rc);
+
+	d_iov_set(&val_iov, NULL, 0);
+	rc = dbtree_lookup(toh, &key_iov, &val_iov);
+	if (rc)
+		D_GOTO(out, rc);
+
+	tmp_root = val_iov.iov_buf;
 
-	memset(&root, 0, sizeof(root));
-	root.root_hdl = DAOS_HDL_INVAL;
 	memset(&uma, 0, sizeof(uma));
 	uma.uma_id = UMEM_CLASS_VMEM;
-
-	rc = dbtree_create_inplace(tree_class, BTR_FEAT_DIRECT_KEY, 32,
-				   &uma, broot, &root.root_hdl);
+	rc = dbtree_create_inplace(DBTREE_CLASS_NV, BTR_FEAT_DIRECT_KEY, 32,
+				   &uma, &tmp_root->btr_root, &tmp_root->root_hdl);
 	if (rc) {
 		D_ERROR("failed to create rebuild tree: "DF_RC"\n", DP_RC(rc));
-		D_FREE(broot);
 		D_GOTO(out, rc);
 	}
 
-	d_iov_set(&val_iov, &root, sizeof(root));
-	rc = dbtree_update(toh, key_iov, &val_iov);
-	if (rc)
-		D_GOTO(out, rc);
-
-	d_iov_set(&val_iov, NULL, 0);
-	rc = dbtree_lookup(toh, key_iov, &val_iov);
-	if (rc)
-		D_GOTO(out, rc);
+	*rootp = tmp_root;
 
-	*rootp = val_iov.iov_buf;
-	D_ASSERT(*rootp != NULL);
 out:
 	if (rc < 0 && daos_handle_is_valid(root.root_hdl))
 		dbtree_destroy(root.root_hdl, NULL);
```

Comment:
delete the inserted entry on error, instead of destroy tree.

Reply:
ah, right